### PR TITLE
Imperial auth

### PR DESCRIPTION
**Pull request by cc-a**
This PR implements an authentication option for Imperial College London with Azure Active Directory. It's a fairly straightforward implementation based off of Orcid minus the whitelisting logic.

We ended up also having to implement the ability to customise the [prompt parameter](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) passed as part of the auth request. This was previously hardcoded as `login+consent` which is a valid value according to the OIDC spec but not for the AAD implementation. I tweaked things so that a different prompt value could be passed through for the Imperial auth but all other providers fall back to the previous default. Extending the ability to customise the prompt value via environment variable could be trivially extended to the other auth providers as well.

How to test:

* In principle could be tested against any Azure AD tenant using details of a registered application.

PR Checklist:

*   [ ] Increase version numbers in `docker-compose.yml`
*   [ ] Link to a GitHub issue
*   [ ] Update documentation
*   [ ] Tests

### CODE REVIEW

**Comment by cc-a** [comment]
I'd like to run this against the CI before writing up as a full pull request. @jmaassen could you please grant the necessary approval for the CI runs?

**Comment by dmijatovic** [comment]
@cc-a All checks are passing. Can we review this PR?

**Comment by cc-a** [comment]
Yes please @dmijatovic

**Comment by ewan-escience** [comment]
The env variable `IMPERIAL_LOGIN_PROMPT` is not yet in the (frontend part of) `docker-compose.yml`, which makes it use the unsupported value `login+consent`.

Furthermore, can you add all the env variables to `deployment/docker-compose.yml` as well?

Is it correct that this implementation can be used as a general Azure AAD implementation? If so, we should rename it as such.

I'm trying to test it out with personally registered app credentials, but Microsoft [changed their policy](https://learn.microsoft.com/en-gb/azure/active-directory/develop/publisher-verification-overview#requirements) and I cannot login yet. I'll see if I can work around it somehow.

**Comment by dmijatovic** [comment]
@cc-a Maybe is useful to join one of our daily stand-ups (at 10 am CET) to discuss our remarks and next steps? 
We can send you invite to our teams session.

**Comment by cc-a** [comment]
Thanks for the feedback @ewan-escience and @dmijatovic. I'll drop by the standup tomorrow morning to discuss.

**Comment by cc-a** [comment]
I've pushed the updates we discussed @dmijatovic @ewan-escience. Seems to be waiting for approval to run the workflows again :thinking:

**Comment by ewan-escience** [comment]
Looks mostly good, just a few small things.

- Can you fill in the value `http://localhost/auth/login/azure` for `AZURE_REDIRECT` in `.env.example`?
- Can you `AZURE` to the allowed values of line 61 of `.env.example`?
- Can you copy the new env variables from `docker-compose.yml` to `deployment/docker-compose.yml`?

Hopefully @dmijatovic can look at the frontent code tomorrow and add the new redirect link to his testing Azure application.

File: .env.example
Comment (dmijatovic) [inline .env.example:1]: Please add AZURE to the list of providers, something like this
```
RSD_AUTH_PROVIDERS=SURFCONEXT;ORCID;AZURE;LOCAL
```

File: frontend/pages/api/fe/auth/azure.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  // extract all props from env and wellknow endpoint
+  const redirectProps = await azureRedirectProps()
+  if (redirectProps) {
+    // create return url and the name to use in login button
+    const redirectUrl = getRedirectUrl(redirectProps)
+    // provide redirectUrl and name/label
+    return {
+      name: process.env.AZURE_DISPLAY_NAME || 'Azure Active Directory',
+      redirectUrl,
+      html: process.env.AZURE_DESCRIPTION_HTML || "Login with your institutional credentials."
```
Comment (dmijatovic) [inline frontend/pages/api/fe/auth/azure.ts:64]: Please use single quotes (') for strings. You can run eslint linter checks locally and fix these automatically :-).

```bash
# run linter locally to check for warnings and errors
yarn lint
# fix automatically warnings
yarn lint --fix
```

**Review by dmijatovic — CHANGES_REQUESTED** [review]
**Excellent work!** I tested with my Azure account and the implementation works properly in my case too. It seems to me generic enough to be used with any Azure AD tenant+app.

There are few cosmetic improvements. I commented on improvement suggestions in the code directly (see my comments). 

I usually run linter and tests locally before I create PR. You can run following commands from then `frontent` folder.

```bash
# run unit tests locally (check if some test are broken by last changes)
yarn test
# run linter locally (check for warnings and errors -> errors are fatal and building of frontend will fail)
yarn lint
```

**Comment by cc-a** [comment]
Awesome. Thanks both. I've made (hopefully) the final tweaks.

**Review by ewan-escience — APPROVED** [review]
Nice work! In principle, it works for me.

When inspecting the database though, I noticed that `home_organisation` was `null`. You try to get the home organisation from `Config.azureOrganisation()`, whish uses the env variable `AZURE_ORGANISATION`, which is neither in the `.env.example` or in the two `docker-compose.yml` files.

I'm approving this PR in case you want to keep it this way for now and fix it later, but also feel free to fix it now. Please let us know what you want to do, so we know when to merge this PR.

**Review by dmijatovic — APPROVED** [review]
Again, excellent work. Thank you for your contribution :+1: 
As Ewan mentioned let us know when this PR can be merged.

**Comment by cc-a** [comment]
Cool. I've added the missing entries for AZURE_ORGANISATION to `.env.example` and the docker compose files and tested. I think this now be ready to merg.

**Pull request merged**